Optional party rank ordering for cities

christosporios opened the pull request:
Solves #128.

Adds an optional party ranking system for ordering people. Cities can enable party-based ordering via the Edit City form, and party admins can set custom member rankings via drag-and-drop. When enabled, the people page sorts by party power (member count) and then by party ranking, and the party page displays members in ranked order.

## Changes

### Data Model:
- Added peopleOrdering enum field to City model (default | partyRank)
- Added rank integer field to Role model (nullable)

### Backend:
- Updated City Edit API to handle peopleOrdering setting
- Created new API endpoint /api/cities/[cityId]/parties/[partyId]/roles/ranking for updating member rankings
- Updated role creation to include rank field

### Frontend:
- Added `peopleOrdering` selector to City Edit form (super admin/city admin only)
- Created drag-and-drop interface (`PartyMemberRankingSheet`) for reordering party members
- Added "Change Member Ordering" button on party pages (only visible when `peopleOrdering === 'partyRank'`)
- Updated sorting logic: **Party page**: sorts members by rank when available, **City people page**: when party ranking enabled, sorts by party power (member count) then by party ranking

### Authorization:
- City peopleOrdering: super admin and city admin only
- Role rank: super admin, city admin, and party admin

### Dependencies:
Added `@dnd-kit/core`, `@dnd-kit/sortable`, and `@dnd-kit/utilities` for drag-and-drop functionality

### Migration
Created migration `20251217134716_add_party_ranking` to add new fields and enum type

This PR also includes fixes for small issues around showing party memberships throughout the app, as the first commit in the PR:
- **Seed data bug**: Added partyId: role.partyId to role creation in seedPersons so roles are linked to parties.
- **Date filtering mismatch**: Updated PartyCard and Party components to use isRoleActive() instead of only checking endDate, ensuring both startDate and endDate are considered.
- **Deduplication bug**: Fixed getPartiesForCity to deduplicate people by ID (like getParty already did), preventing people with multiple roles from being counted multiple times.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables party-based ordering of people with persistent, per-party member rankings.
> 
> - Data: Adds `City.peopleOrdering` enum (`default` | `partyRank`) and `Role.rank` (nullable); migration `20251217134716_add_party_ranking`
> - Backend APIs: 
>   - Updates city create/edit (`/api/cities`, `/api/cities/[cityId]`) to read/write `peopleOrdering`
>   - New `POST /api/cities/[cityId]/parties/[partyId]/roles/ranking` to update `Role.rank` with validation and cache revalidation
> - Sorting/Utils: New `@/lib/sorting/people` (party grouping, rank-aware sorting), `getActiveRoleCondition`, `getActivePartyRole`; replaces ad-hoc checks across app
> - Frontend:
>   - City edit form: new `peopleOrdering` selector (i18n strings)
>   - Party page: "Change Member Ordering" opens `PartyMemberRankingSheet` (DnD via `@dnd-kit/*`); members sorted by rank; labels i18n-ized
>   - City people page: passes `city`; uses `sortPeople` to sort by party power then rank when enabled
> - DB/Types: Centralizes shared Prisma types under `@/lib/db/types` (e.g., `RoleWithRelations`, `roleWithRelationsInclude`); updates people/parties/segments queries to use them and active-role filtering
> - Seed fix: include `partyId` when seeding roles
> - Deps: adds `@dnd-kit/core`, `@dnd-kit/sortable`, `@dnd-kit/utilities`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8f0a08915218e5e75a5a4149c9b98d257d0ee584. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

christosporios committed:
Seed data party membership bug; date filtering mismatch and deduplication mismatch between party cards and party pages

christosporios committed:
Party Rank Ordering for people

kouloumos commented on src/components/cities/CityForm.tsx:
```diff
@@ -252,6 +257,32 @@ export default function CityForm({ city, cityMessage, onSuccess }: CityFormProps
+                            </FormItem>
+                        )}
+                    />
+                )}
```
The rest of this page uses translations, please add here as well.

kouloumos commented on src/lib/db/parties.ts:
```diff
@@ -70,30 +70,24 @@ export async function getParty(id: string): Promise<PartyWithPersons | null> {
-                        ]
                     }
+                    // Removed the where clause to include all roles (active and inactive)
+                    // The component will filter them appropriately
```
Can you clarify why this is needed and what the benefit is of doing this filtering on the client side?

My concern is that inactive people aren’t really part of the Party anymore, so returning them at all feels a bit misleading. Would it make sense to model this explicitly with a flag instead?

Also, applying this change to `getParty` but not to `getPartiesForCity` could lead to inconsistent behavior and potential confusion down the line. It might be worth aligning the two APIs or documenting the difference.

kouloumos commented on src/lib/db/parties.ts:
```diff
@@ -149,11 +143,19 @@ export async function getPartiesForCity(cityId: string): Promise<PartyWithPerson
+        // Deduplicate people by id since a person can have multiple roles
         const partiesWithPeople = parties.map(party => {
             const { roles, ...partyWithoutRoles } = party;
+            const peopleMap = new Map<string, Person & { roles: (Role & { city?: City | null; administrativeBody?: AdministrativeBody | null; party?: Party | null })[] }>();
```
We should avoid redefining types in multiple places.

kouloumos committed:
refactor(db): optimize PartyWithPersons type using Prisma.PartyGetPayload

kouloumos committed:
refactor(db): extract normalizePartyPeople helper to deduplicate party members

kouloumos commented on src/components/cities/CityPeople.tsx:
```diff
@@ -32,7 +35,12 @@ export default function CityPeople({
+            return sortPeopleByPartyRank(allPeople, partiesWithPersons, city.peopleOrdering);
+        }
+        return sortPersonsByLastName(allPeople);
+    }, [allPeople, partiesWithPersons, city?.peopleOrdering]);
```
`sortPeopleByPartyRank` already defaults to `sortPersonsByLastName` when `peopleOrdering !== 'partyRank'`. Why not rename it to `sortPeople` and use just that without the need for extra clauses?

kouloumos commented on src/components/parties/Party.tsx:
```diff
@@ -72,10 +73,30 @@ function PartyMembersTab({
+                                size="sm"
+                                onClick={() => setIsRankingSheetOpen(true)}
+                            >
+                                Change Member Ordering
```
needs localization to match the rest of the page

kouloumos reviewed (commented):
Looks great overall -- especially the drag-and-drop interface. I’ve tested locally. I left a few inline comments.

One question regarding the sorting logic: with the new ordering, should we preserve the default behavior where the head of the party is always shown first? We currently do this on the individual party pages, but not on the main city page.

Based on user feedback, when people visit a city’s “People” page, they generally expect to see the mayor listed first. It might be worth considering a default ordering where at least the head of the party — and possibly the Αντιδήμαρχοι as well — are prioritized at the top.

I've gone ahead and made some changes to this code based on my comments. Feel free to cherry-pick https://github.com/schemalabz/opencouncil/commit/ae166134dc54490039a49465a45852a9ad3302b1 and https://github.com/schemalabz/opencouncil/commit/dd3876fff3d2c7951af73f131d7a1f44fb7378e7 from [pr-137-fixes](https://github.com/kouloumos/opencouncil/commits/pr-137-fixes/)

christosporios committed:
move role ranking to a separate file; correctly validate user can edit

christosporios committed:
check role activity in sortPeopleByPartyRank

kouloumos committed:
refactor(db): introduce getActiveRoleCondition for streamlined role filtering

kouloumos committed:
refactor(types): centralize RoleWithRelations and roleWithRelationsInclude for improved type management

kouloumos committed:
refactor: move people sorting into src/lib/sorting/people.ts

- Created `src/lib/sorting/people.ts` with 6 reusable sorting/grouping functions
- Replaced inline sorting logic in PartyCard.tsx and chat route
- Updated imports across 6 files to use centralized module
- Deleted `src/components/utils.ts`
